onnxruntime/core/providers/coreml/builders/impl/clip_op_builder.cc

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/coreml/builders/model_builder.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+#include "core/providers/shared/utils/utils.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ClipOpBuilder : public BaseOpBuilder {
+  // Add operator related
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+void ClipOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // Both min and max will be inject in the layer, no need to add to the model
+  if (node.SinceVersion() >= 11) {
+    if (node.InputDefs().size() > 1)
+      model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
+
+    if (node.InputDefs().size() > 2)
+      model_builder.AddInitializerToSkip(node.InputDefs()[2]->Name());
+  }
+}
+
+Status ClipOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                            const Node& node,
+                                            const logging::Logger& logger) const {
+  const auto& node_name = node.Name();
+  const auto& input_name = node.InputDefs()[0]->Name();
+  const auto& output_name = node.OutputDefs()[0]->Name();
+  float min, max;
+  ORT_RETURN_IF_NOT(GetClipMinMax(model_builder.GetInitializerTensors(), node, min, max, logger), "GetClipMinMax failed");
+
+  bool has_min = min != std::numeric_limits<float>::lowest();
+  bool has_max = max != std::numeric_limits<float>::max();
+
+  if (!has_min && !has_max) {
+    // Clip without min/max is an identity node
+    // In CoreML we don't have identity, use ActivationLinear instead
+    std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+    layer->mutable_activation()->mutable_linear()->set_alpha(1.0f);
```

Comment:
Just curious about how we use `ActivationLinear` to replace identity node here? For example, what does the `set_alpha=1.0f` do? I thought the linear activation function is \[f(x) = \alpha x + \beta\] and `alpha` and `beta` are initially set to be 1 and 2.

Reply:
For `alpha` and `beta` shown below
https://github.com/apple/coremltools/blob/82c8d96cb396cc19cafc2da0ed34c1022543b486/mlmodel/format/NeuralNetwork.proto#L290-L303

The numbers after the fields are not the default values
https://developers.google.com/protocol-buffers/docs/proto#assigning_field_numbers